### ISSUE TITLE
How to handle moustache app macro

### ISSUE THREAD

**Issue by agriffis** [kkinnear/zprint#56]
What do you recommend to handle a DSL like the [moustache app macro](https://github.com/cgrand/moustache)?

The macro has four usages:

1. wrap a Ring handler,
2. to define routes,
3. to dispatch on HTTP methods
4. and to render plain text

Usage 1 (wrapping handler with middleware) is probably `:flow-body`

Usages 2 and 3 (routes and methods) are `:pair-fn`

Usage 4 is probably `:hang-body`

**Comment by kkinnear** [comment]
I don't have personal experience with moustache, so I've only looked at the walkthrough in the readme.  I've experimented with formatting the gist of the walkthrough, and I think that defining `app` as `:pair-fn` makes the most sense.  If you are using lein-zprint, you can then put a `;!zprint` directive prior to any cases where `:pair-fn` doesn't work well to define `app` as something else.   That said, defining `app` as `:pair-fn` formatted the gist of the walkthrough at least as well (or in some cases better) than it was formatted by hand, so I don't think you are going to go far wrong with  that approach.

**Comment by agriffis** [comment]
Okay, thanks Kim! I haven't used lein-zprint so I'll look into that.

**Comment by agriffis** [comment]
@kkinnear I think to do this right I need a variant of `options-fn` that works for `fn-map` instead of just vectors

Consider a few examples:

```
(m/app [route] handler
       [route] handler
       [route] handler)
```

That works fine with `:pair-fn`. Similarly dispatching on HTTP methods:

```
(m/app :get  (m/app [route] handler
                    [route] handler)
       :post (m/app [route] handler
                    [route] handler))
```

However as soon as we throw a middleware in there, it's all confused:

```
(m/app middle1 [route]
       handler [route]
       handler)
```

and there can be an arbitrary number of middlewares:

```
(m/app middle1 middle2
       middle3 middle4
       middle5 [route]
       handler [route]
       handler)
```

Ideally that last example would be one of these instead:

<table><tr><td valign=top>

```
(m/app middle1
       middle2
       middle3
       middle4
       middle5
       [route] handler
       [route] handler)
```

</td><td valign=top>

```
(m/app middle1 middle2 middle3 middle4 middle5
       [route] handler
       [route] handler)
```

</td></tr></table>

I think in order to achieve this, I would need to be able to inspect the params to set the appropriate function classification. Additionally, it would be helpful to be able to use an arbitrary arg number, for example `:arg5-pair` (which shouldn't really be that hard if you want to make it completely dynamic, for unknown function classifications, you could convert to a string, match against a pattern and extract the number)

What do you think?

**Comment by agriffis** [comment]
By the way, if you wouldn't mind re-opening this issue, I'd appreciate that, so it doesn't get lost

**Comment by kkinnear** [comment]
Thanks for the examples.  I see your point.  The "ideally the last example would be one of these" versions look a great deal like [constant pairing](https://github.com/kkinnear/zprint#constant-pair-true), where I look backwards from the end of a list, and if there are constants in positions n-1, n-3, etc., I pair up those last things.  

One approach would be to allow you to configure your own definition of "constant", which right now is a number, string, or keyword.  Then for `m/app` you could redefine constant inside that form in any way you want.  Probably a vector, which would allow the routes to pair up with the handlers.

A more general approach would be as you suggested, to support `options-fn` for lists.  For performance, you would only define one for the top level `m/app`, I think.  

The hardest part of what you suggest would actually be the `:arg-n` generalization.  It sounds easy, but I suspect that it isn't -- not that I've tried to implement it yet.  It is challenging in part because things like comments can occur in between arguments, in part because the "arguments" can themselves be arbitrarily complex expressions which may not fit well on the current line.

It might be better to allow an `options-fn` to specify the number of things that should be paired up at the end of a list (like constant pairing, but without the checks for constants).  Then the rest of the list would end up like the left-hand of your two suggested outcomes. (Impressive use of markdown to get the two examples side by side, I didn't know you could do that!)

Anyway, I've written all of this down so that when I get back to it, I can recover my context.  I have several things to implement first, and the time to do them is in moderately short supply this time of year.  I will put this on the list of things to do and work on it as soon as I can release the current things. Thank you for the suggestion!

**Comment by agriffis** [comment]
Oh, constant pairing looks interesting. No matter how many times I've been through the doc, there's always something I haven't seen yet.

> Then for m/app you could redefine constant inside that form in any way you want. Probably a vector, which would allow the routes to pair up with the handlers.
> ...
> It might be better to allow an options-fn to specify the number of things that should be paired up at the end of a list (like constant pairing, but without the checks for constants). Then the rest of the list would end up like the left-hand of your two suggested outcomes.

Either option sounds good to me! Thanks for considering it.

**Comment by agriffis** [comment]
@kkinnear Just wanted to check back, has anything changed in zprint since last year, that would make it easier to configure the definition of "constant" for constant pairing? Rereading this issue, it seems like if I could treat vectors as constants, then I could really get the `m/app` macro to format properly.

**Comment by kkinnear** [comment]
Thanks for the reminder.  I'm embarrassed to say that this slipped off my radar despite being an open issue.  I'm really sorry that I didn't address this sooner!  The good news is that the only important thing I'm working on right now is how to handle Windows line endings, so something more creative -- like this issue -- would be most welcome.  I'll see what I can do and let you know, sooner than 10 months this time.  I'm really glad we had this conversation late last year and I wrote it down so I don't have to start from scratch!  I'm excited to work on this!

**Comment by kkinnear** [comment]
So, here is what I've prototyped so far to solve this issue.  I'd like your feedback if this would actually work for you before I commit into the next release.

This approach allows you to supply a function to decide what is a "constant", such that constant-pairing will pair up things from the end of a list for as far up (back) as there are "constants" for every other element.  

This first example says that vectors are constants, so they are paired up.  However, this also makes the test (which I added, and isn't something that you would do, of course) for standard constant pairing fail:

```clojure
 (czprint mapp1 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-fn vector?}}]}})
(m/app middle1
       middle2
       middle3
       (this is
             "a"
             test
             "this"
             is
             "only a"
             test)
       middle4
       middle5
       [route] handler
       [route] handler)
```

Here is the same approach, but I've fixed up the options map in the `:fn-map` to only consider vectors as "constants" for the top level of `m/app`, and then go back to normal constant pairing:

```clojure
(czprint mapp1 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-fn vector?} :next-inner {:list {:constant-fn nil}}}]}})
(m/app middle1
       middle2
       middle3
       (this is
             "a" test
             "this" is
             "only a" test)
       middle4
       middle5
       [route] handler
       [route] handler)
```
This works for `:constant-fn` functions that you write too:
```
(czprint mapp {:parse-string? true :width 40 :list {:constant-fn #(vector? %)}})
(m/app middle1
       middle2
       middle3
       middle4
       middle5
       [route] handler
       [route] handler)
```
To generalize a bit for the dispatching on http methods:

```clojure
(czprint mapp2 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-fn #(or (vector? %) (keyword? %))}}]}})
(m/app :get (m/app [route] handler
                   [route] handler)
       :post (m/app [route] handler
                    [route] handler))
```
To turn it off so that the handlers don't get funny constant pairing:

```clojure
(czprint mapp2 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-fn #(or (vector? %) (keyword? %))} :next-inner {:list {:constant-fn nil}}}]}})
(m/app :get (m/app [route] handler
                   [route] handler)
       :post (m/app [route] handler
                    [route] handler))
```
A more complex example with both handlers and http dispatching:
```clojure
(czprint mapp3 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-fn #(or (vector? %) (keyword? %))} :next-inner {:list {:constant-fn nil}}}]}})
(m/app :get (m/app middle1
                   middle2
                   middle3
                   [route] handler
                   [route] handler)
       :post (m/app middle
                    of
                    the
                    road
                    [route] handler
                    [route] handler))
```
The standard constant-pairing requires two things before it kicks in, which all of your examples have, but here is what is necessary to make the pairing work for even a single pair:
```clojure
(czprint mapp4 {:parse-string? true :width 40 :fn-map {"app" [:none {:list {:constant-pair-min 1 :constant-fn #(or (vector? %) (keyword? %))} :next-inner {:list {:constant-fn nil :constant-pair-min 2}}}]}})
(m/app :get (m/app middle1
                   middle2
                   middle3
                   [route] handler
                   [route] handler))
```

To me, this looks like it would solve the issue that you have raised.

But this only works to pair things up if you have vectors (or something identifiable as a "constant" by a function that you supply) as the n-1, n-3, ... elements of the list with `m/app`.   This works for the examples that you have given, above.

The question is -- would this work in the more general case for you?

**Comment by agriffis** [comment]
Wow Kim! Pretty sure this is perfect. Can't wait to take it for a spin. This all works in `.zprintrc`, right?

**Comment by kkinnear** [comment]
For sure, this will work in `.zprintrc`.  I'm also planning to define a `:style` which includes it, currently named `:moustache`.

Here is the current `:style` that I'm working with.  I ended up extending the existing `:constant-pair` constant definitions
by adding vectors, instead of going for just keywords and vectors. 

```clojure
      :moustache {:fn-map {"app" [:none
                                  {:list {:constant-pair-min 1,
                                          :constant-pair-fn #(or (keyword? %)
                                                                 (string? %)
                                                                 (number? %)
                                                                 (= true %)
                                                                 (= false %)
                                                                 (vector? %))},
                                   :next-inner {:list {:constant-pair-min 4,
                                                       :constant-pair-fn
                                                         nil}}}]}},
```
It seemed like that might be a touch more generally useful. I think that will still give you what you want as well.

**Comment by agriffis** [comment]
Looks great to me. When can I try it? :smile:

**Comment by kkinnear** [comment]
Need to finish the line endings work, but that is it for this release.  I'm thinking that I will do 1.0.2 fairly soon, possibly this coming weekend, but more likely the next weekend.

**Comment by kkinnear** [comment]
This is now available in 1.0.2.  See the `:moustache` style, as well as the pointer to the documentation for "more flexible constant pairing" in the "recent additions" section of the readme (right at the beginning).

**Comment by agriffis** [comment]
@kkinnear Thank you! This is great. Thanks again for all your work on this.

I'm just stumbling on one thing, now... If I want to combine this with `{:justify? true}`, just for the `app` macro, is that hard to do? I've tried a few things, and the only way I've succeeded is `:pair {:justify? true}` in `.zprintrc`, but then it applies everywhere:

```clojure
{:pair {:justify? true}
 :style [:moustache :respect-bl]}
```

I was able to make it work with the file comment API, but I'm pretty sure I'd like this to be the default for the `app` macro in this project:

```clojure
;!zprint {:format :next :pair {:justify? true}}
```

If I try to add the style to my `$PROJECT/.zprint.edn` then zprint ignores my configuration entirely:

```clojure
{:style [:moustache :respect-bl]
 :style-map
   {:moustache
      {:fn-map {"app" [:none
                       {:list {:constant-pair-min 1
                               :constant-pair-fn #(or (keyword? %)
                                                      (string? %)
                                                      (number? %)
                                                      (= true %)
                                                      (= false %)
                                                      (vector? %))}
                        :pair {:justify? true}
                        :next-inner
                          {:list {:constant-pair-min 4 :constant-pair-fn nil}
                           :pair {:justify? false}}}]}}}}
```

What would you suggest?

**Comment by agriffis** [comment]
If this is hard, please don't worry about it. It might be enough to use the file comment API (which wouldn't have sufficed for the rest of this issue)

**Comment by kkinnear** [comment]
I'm giving this a look now, I'll let you know what I figure out.

**Comment by kkinnear** [comment]
This is a really tricky one.  You contributed to the issue with fn's in .zprintrc files, #124, and when I "fixed" that, I left out the graalVM binaries (for reasons that I haven't had a chance to review).  So, that's why when you try to redefine the `:constant-pair-fn` it isn't going to work if you are using the pre-built binaries.  Period.  If you are using `lein zprint`, then the `#( ...)` function construction won't work, but the `(fn [x] (or ...))` construction might well work fine.  I didn't test that because I was assuming that you were using the pre-build binaries.

I will review why I sidestepped fixing the graalVM binaries when loading functions in the next day or so, but whatever I find out isn't going to solve your problem in the short term.

Another thing that (if you could redefine the functions) would be getting you is:

> Note that styles are applied before the rest of the elements of a options map, so that you can override elements of the style that you wish to change by specifying an explicit element in the options map.

so that redefining a style and then trying to apply (i.e., use) it is problematic (i.e, doesn't work).

So, while I don't have a solution to these problems, I have something that is better than nothing, but not perfect (and I'm not entirely sure why at this juncture, actually).

If you do this in your `~/.zprintrc`:
```
{:fn-map {"app" [:none
                 {:next-inner {:list
                                 {:constant-pair-fn nil, :constant-pair-min 4},
                               :pair {:justify? false}},
                  :style [:moustache :justified]}]}}
```
you will get justification at the top level use of `app`, but not on any nested invocations of `app`.  

If you do this:
```
{:fn-map {"app" [:none {:style [:moustache :justified]}]}}
```
you will get justification in all of the uses of `app`, but also in all of the stuff inside of the uses of `app`, which may not be what you want.

At this stage, I don't have a way to get justification only in the top levels of `app`, though your re-definition of `:moustache` should work fine if it could be read-in.  It seems to work fine when I try it at the repl, but there is no way for you to input it using the graalVM binaries.

I will look at these things:

1. Why I didn't fix the graalVM binaries to be able to load functions.
2. Why doesn't the first `~/.zprintrc` example above do what I expect? (That is, justify all of the top levels of `app`, and nothing else.)
3. Augment the definition of the`:moustache` style with `{:pair {:justifiy? true}}` and then turn it off with `:next-inner`, since that is actually what you wanted in the first place.
4. See if I can figure out how to redefine a style and then apply it and have that work in the .zprintrc.

I could probably do 3, above, in the next two days if that is important to you.   There is an overhead to every release in testing, the steps of which are automated but the overall process is not, so it is a couple of hours to release a new version even when I've just done it.  Which I will do if it would make things a lot better for you.  Otherwise, I'll try to figure out the larger picture and fix what I can, and then extend the `-w` switch, and then release all of that together (which is probably a couple of weeks, I would guess).

**Comment by kkinnear** [comment]
My current thinking is to move to `sci` for loading the config option maps, which will define functions with either `(fn [] ...)` or `#(... %)`, and execute them --theoretically even with graalVM.  Which seems like a win.  I need to try it, with graalVM for sure, to know if this is really a solution.  But it looks promising from a first try at the REPL.

**Comment by agriffis** [comment]
I wondered if this was related to #124. I hadn't heard of it before, but `sci` looks like a good fit!

> 3. Augment the definition of the `:moustache` style with `{:pair {:justify? true}}` and then turn it off with `:next-inner`, since that is actually what you wanted in the first place.

I think that makes sense. I'm sorry I didn't realize I wanted this earlier!

> I could probably do 3, above, in the next two days if that is important to you.

That's a kind offer, but no need to rush. This is already usable for me as-is, and I'll just take advantage of the update when it's available.

**Comment by kkinnear** [comment]
So, I did `sci` for loading option maps.  But all of that should not matter, now, since I have (finally) released 1.1.0.  This has the `:mustache` style that I think you want built-in.  I think it does what you want.  If not, let me know.  Given that I've restructured not only the actual `:mustache` style by splitting it into two parts, but also the way that one style can employ another, and that you can define and use a style in the same `.zprintrc`, it should be relatively straightforward to get you the style you want even if the built-in one is still not quite right.

I've put a lot of effort into making all of this work, so if it isn't (yet) quite right, please let me know!

Thanks for your patience, and for also requesting an interesting capability.  As you can see, it has caused me to clean up a number of issues around styles and option maps, which was really good to do!

**Comment by agriffis** [comment]
I just took it for a spin, Kim, and it looks amazing. Thank you!!
